### PR TITLE
add Exiv2

### PR DESCRIPTION
**Pull request by Crghilardi**
Add a build_tarballs.jl for the [Exiv2](https://exiv2.org/) library

Tested locally on `linux-gnu`, `linux-musl` and `mingw` with no issues.

### CODE REVIEW

File: E/Exiv2/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    LibraryProduct("libexiv2", :libexiv2),
+    ExecutableProduct("exiv2", :exiv2)
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency(PackageSpec(name="Zlib_jll", uuid="83775a58-1f1d-513f-b197-d71354ab007a"))
+    Dependency(PackageSpec(name="Expat_jll", uuid="2e619515-83b5-522b-bb60-26c02a35a201"))
+    Dependency(PackageSpec(name="libpng_jll", uuid="b53b4c65-9356-5827-b1ea-8c7a1a84506f"))
+    Dependency(PackageSpec(name="Gettext_jll", uuid="78b55507-aeef-58d4-861c-77aaff3498b1"))
```
Comment (giordano) [inline E/Exiv2/build_tarballs.jl:53]: Is gettext really needed at runtime?

File: E/Exiv2/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    Dependency(PackageSpec(name="Libiconv_jll", uuid="94ce4f54-9a6c-5748-9c1c-f9c7231a4531"))
+]
+
+# Build the tarballs, and possibly a `build.jl` as well.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; julia_compat="1.6")
```
Comment (giordano) [inline E/Exiv2/build_tarballs.jl:58]:
```suggestion
build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; julia_compat="1.6")

```

File: E/Exiv2/build_tarballs.jl (hunk 1 above)
Comment (Crghilardi) [reply E/Exiv2/build_tarballs.jl:53]: I don't think so? I added it following the [dependency list](https://github.com/exiv2/exiv2#24-dependencies) on the `README`. Looking at the top level `CMake` file again now, the flag that uses that is off by default, so we can probably get rid of it or comment it out?

Comment (giordano) [reply E/Exiv2/build_tarballs.jl:53]: Gettext is an annoying dependency, if it can be avoided, much better

**Comment by Crghilardi** [comment]
I don't quite understand what is going on with the `i686-linux-musl-cxx11` check, especially when the `i686-linux-musl-cxx03` one passes?

```
[19:28:21] CMakeFiles/exiv2lib.dir/basicio.cpp.o: In function `Exiv2::FileIo::write(Exiv2::BasicIo&) [clone .part.25] [clone .constprop.97]':
[19:28:21] basicio.cpp:(.text+0xa64): undefined reference to `__stack_chk_fail_local'
[19:28:21] CMakeFiles/exiv2lib.dir/basicio.cpp.o: In function `Exiv2::RemoteIo::path[abi:cxx11]() const':
[19:28:21] basicio.cpp:(.text+0xc44): undefined reference to `__stack_chk_fail_local'
[19:28:21] CMakeFiles/exiv2lib.dir/basicio.cpp.o: In function `Exiv2::FileIo::path[abi:cxx11]() const [clone .localalias.80]':
[19:28:21] basicio.cpp:(.text+0xd14): undefined reference to `__stack_chk_fail_local'
[19:28:21] CMakeFiles/exiv2lib.dir/basicio.cpp.o: In function `Exiv2::FileIo::Impl::Impl(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
[19:28:21] basicio.cpp:(.text+0xf75): undefined reference to `__stack_chk_fail_local'
[19:28:21] CMakeFiles/exiv2lib.dir/basicio.cpp.o: In function `Exiv2::FileIo::Impl::stat(Exiv2::FileIo::Impl::StructStat&) const':
[19:28:21] basicio.cpp:(.text+0x165c): undefined reference to `__stack_chk_fail_local'
[19:28:21] CMakeFiles/exiv2lib.dir/basicio.cpp.o:basicio.cpp:(.text+0x16f9): more undefined references to `__stack_chk_fail_local' follow
[19:28:21] collect2: error: ld returned 1 exit status

```

I tried adding a `preferred_gcc` version with 5,6,and 7 unsuccessfully. I have also tried adding a [-DBUILD_WITH_STACK_PROTECTOR](https://github.com/Exiv2/exiv2/pull/1756) flag with the same result.

 The `0.27.4` release is listed as 499 commit behind master so I haven't been able to find any patches in master worth pulling out yet.

**Comment by giordano** [comment]
> I don't quite understand what is going on with the i686-linux-musl-cxx11 check,

We've got the same issues with `__stack_chk_fail_local` in #845 and #3249,, in both cases I think we ended up not going through that code path

> especially when the i686-linux-musl-cxx03 one passes?

That uses GCC 4.8, the C++11 string ABI necessarily requires GCC 5 (well, 4.9 but we don't have it), maybe that's relevant

**Comment by Crghilardi** [comment]
So the issue came down to the addition of a `-fstack-protector-strong` flag in `i686-linux-musl-cxx11` build from a series of `check_cxx_compiler_flag` statements that was not included in the cxx03 build. Apparently, gcc emits slightly different code for 32 bit stack stuff and musl takes a slightly different stance on how this should all be handled.



`i686-linux-musl-cxx11` CMake output

```

[19:28:14] -- Performing Test HAS_FSTACK_CLASH_PROTECTION
[19:28:14] -- Performing Test HAS_FSTACK_CLASH_PROTECTION - Failed
[19:28:14] -- Performing Test HAS_FCF_PROTECTION
[19:28:14] -- Performing Test HAS_FCF_PROTECTION - Failed
[19:28:14] -- Performing Test HAS_FSTACK_PROTECTOR_STRONG
[19:28:15] -- Performing Test HAS_FSTACK_PROTECTOR_STRONG - Success


[19:28:16] -- CMAKE_CXX_STANDARD:
[19:28:16] --  --- Compiler flags --- 
[19:28:16] -- General:           
[19:28:16] 	 -fstack-protector-strong
[19:28:16] 	 -Wp,-D_GLIBCXX_ASSERTIONS
[19:28:16] 	 -Wp,-D_FORTIFY_SOURCE=2
[19:28:16] 	 -Wall
[19:28:16] 	 -Wcast-align
[19:28:16] 	 -Wpointer-arith
[19:28:16] 	 -Wformat-security
[19:28:16] 	 -Wmissing-format-attribute
[19:28:16] 	 -Woverloaded-virtual
[19:28:16] 	 -W


`i686-linux-musl-cxx03` CMake output

```

[19:27:18] -- Performing Test HAS_FSTACK_CLASH_PROTECTION
[19:27:18] -- Performing Test HAS_FSTACK_CLASH_PROTECTION - Failed
[19:27:18] -- Performing Test HAS_FCF_PROTECTION
[19:27:18] -- Performing Test HAS_FCF_PROTECTION - Failed
[19:27:18] -- Performing Test HAS_FSTACK_PROTECTOR_STRONG
[19:27:18] -- Performing Test HAS_FSTACK_PROTECTOR_STRONG - Failed

[19:27:20] -- CMake Generator:   Unix Makefiles
[19:27:20] -- CMAKE_BUILD_TYPE:  Release
[19:27:20] -- Compiler info: GNU (/opt/bin/i686-linux-musl-libgfortran3-cxx03/i686-linux-musl-g++) ; version: 4.8.5
[19:27:20] -- CMAKE_CXX_STANDARD:
[19:27:20] --  --- Compiler flags --- 
[19:27:20] -- General:           
[19:27:20] 	 -Wp,-D_GLIBCXX_ASSERTIONS
[19:27:20] 	 -Wp,-D_FORTIFY_SOURCE=2
[19:27:20] 	 -Wall
[19:27:20] 	 -Wcast-align
[19:27:20] 	 -Wpointer-arith
[19:27:20] 	 -Wformat-security
[19:27:20] 	 -Wmissing-format-attribute
[19:27:20] 	 -Woverloaded-virtual
[19:27:20] 	 -W
```

**Comment by Crghilardi** [comment]
@giordano I was reading [in this thread](https://www.openwall.com/lists/musl/2018/09/11/2) that this might be handled in some form by `libssp` (or -libssp_nonshared?). It looks like that gets [disabled in the BB musl builds?](https://github.com/JuliaPackaging/Yggdrasil/blob/03730333e7b687f1eec50f30a9856e68f922824b/0_RootFS/gcc_common.jl#L337). Am I understanding that correctly and we are disabling `libssp` for BB?

File: E/Exiv2/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+# The products that we will ensure are always built
+products = [
+    LibraryProduct("libexiv2", :libexiv2),
+    ExecutableProduct("exiv2", :exiv2)
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency(PackageSpec(name="Zlib_jll", uuid="83775a58-1f1d-513f-b197-d71354ab007a"))
+    Dependency(PackageSpec(name="Expat_jll", uuid="2e619515-83b5-522b-bb60-26c02a35a201"))
```
Comment (giordano) [inline E/Exiv2/build_tarballs.jl:60]: See #3024
```suggestion
    Dependency("Expat_jll"; compat="2.2.10")
```

**Review by giordano — COMMENTED** [review]
Thanks for the great detective work!  I think this is a good solution for the time being, also because i686-linux-musl is the only platform for which Julia doesn't really work, so not many people are going to complain anyway if this breaks anything :smile:

**Pull request merged**